frontend/src/metabase/visualizations/echarts/cartesian/model/dataset.ts

```diff
@@ -177,9 +178,118 @@ export const getJoinedCardsDataset = (
+  dataset: ChartDataset,
+) => Record<DataKey, RowValue>;
+
+type ConditionalTransform = {
+  condition: boolean;
+  fn: TransformFn;
+};
+
+export const transformDataset = (
+  dataset: ChartDataset,
+  transforms: (TransformFn | ConditionalTransform)[],
+): ChartDataset => {
+  // Filter out transforms that don't apply
+  const effectiveTransforms = transforms
+    .map(transform => {
+      if (typeof transform === "function") {
+        return transform;
+      } else if (transform.condition) {
+        return transform.fn;
+      } else {
+        return null;
+      }
+    })
+    .filter(isNotNull);
+
+  // Apply the filtered transforms
+  return dataset.map((record, index) => {
+    return effectiveTransforms.reduce((acc, transform) => {
+      return transform(acc, index, dataset);
+    }, record);
+  });
+};
+
+const getNumberOrZero = (value: RowValue): number =>
+  typeof value === "number" ? value : 0;
+
+export const computeTotal = (
+  row: Record<DataKey, RowValue>,
+  keys: DataKey[],
+): number => keys.reduce((total, key) => total + getNumberOrZero(row[key]), 0);
+
+export const getNormalizedDatasetTransform = (
+  seriesDataKeys: DataKey[],
+  dimensionKey: DataKey,
+): TransformFn => {
+  return datum => {
+    const total = computeTotal(datum, seriesDataKeys);
+
+    // Copy the dimension value
+    const normalizedDatum: Record<DataKey, RowValue> = {
+      [dimensionKey]: datum[dimensionKey],
+    };
+
+    // Compute normalized values for metrics
+    return seriesDataKeys.reduce((acc, key) => {
+      const numericValue = getNumberOrZero(datum[key]);
+      acc[key] = numericValue / total;
+      return acc;
+    }, normalizedDatum);
+  };
+};
+
+export const getKeyBasedDatasetTransform = (
+  keys: DataKey[],
+  valueTransform: (value: RowValue) => RowValue,
+): TransformFn => {
+  return datum => {
+    const transformedRecord = { ...datum };
+    for (const key of keys) {
+      if (key in datum) {
+        transformedRecord[key] = valueTransform(datum[key]);
+      }
+    }
+    return transformedRecord;
+  };
+};
+
+export const getNullReplacerTransform = (
+  settings: ComputedVisualizationSettings,
+  seriesModels: SeriesModel[],
+): TransformFn => {
+  const replaceNullsWithZeroDataKeys = seriesModels.reduce(
+    (seriesDataKeys, seriesModel) => {
+      const shouldReplaceNullsWithZeros =
+        settings.series(seriesModel.legacySeriesSettingsObjectKey)[
+          "line.missing"
+        ] === "zero";
+
+      if (shouldReplaceNullsWithZeros) {
+        seriesDataKeys.push(seriesModel.dataKey);
+      }
+
+      return seriesDataKeys;
+    },
+    [] as DataKey[],
+  );
+
+  return getKeyBasedDatasetTransform(
+    replaceNullsWithZeroDataKeys,
+    (value: RowValue) => {
+      return value === null ? 0 : value;
+    },
+  );
+};
+
 export const applySquareRootScaling = (value: RowValue): RowValue => {
   if (typeof value === "number") {
-    return Math.sqrt(value);
+    const sign = value > 0 ? 1 : -1;
```

Comment:
I think this was copied from another PR but it should be fine since it won't conflict